### DISCUSSION TITLE
Learning internal details of Ohm

### DISCUSSION THREAD

**Discussion by MuhammedZakir** [ohmjs/ohm#311] [General]
Thank for Ohm and the JS implementation. Nicely done! :+1: :heart:

I want to learn the internal/implementation details. Do you have any suggestion/advice on where I should start? I will look through the code. But asked here in case I get something nice. :-)

**Comment by pdubroy** [comment]
Hi @MuhammedZakir, sorry I haven't had a chance to respond until now — have you taken a look at the code already? If so, let me know if you have any questions.

To understand how a grammar (e.g. in a .ohm file) is transformed from text into a Grammar instance, you dig into [compileAndLoad in main.js](https://sourcegraph.com/github.com/harc/ohm@e8857899d7964d1063a26c3d3aba6f69265d6791/-/blob/packages/ohm-js/src/main.js?L257:10). To learn how matching works, you can look at the [Matcher class](https://sourcegraph.com/github.com/harc/ohm@e885789/-/blob/packages/ohm-js/src/Matcher.js). 

Another thing worth checking out is the artifacts from our "Incremental Packrat Parsing" paper. For that paper, we implemented [a very simple Ohm-like tool in about 150 LoC](https://github.com/ohmlang/sle17/blob/master/src/standard.js). It's missing many of the features that Ohm has (including support for left recursion) but it might be a good place to start to understand the basic concepts.
